placement/cluster_mapping_provider.go

```diff
@@ -0,0 +1,476 @@
+				readCutoverTime:  m.readCutoverTime,
+				writeCutoverTime: m.writeCutoverTime,
+				prior:            prior,
+			},
+			shards: m.shards.Clone(),
+		}
+	}
+
+	return clone
+}
+
+// update updates the database with new mapping information
+func (dbm *databaseMappings) update(db *schema.Database) error {
+	// Short circuit if we are already at the proper version
+	if dbm.version == db.Version {
+		return nil
+	}
+
+	// Apply new rules
+	// NB(mmihic): Assumes mapping rule sets are sorted in ascending version order
+	for _, rules := range db.MappingRules {
+		if rules.ForVersion <= dbm.version {
+			continue
+		}
+
+		if err := dbm.applyRules(rules); err != nil {
+			return err
+		}
+	}
+
+	dbm.version = db.Version
+	return nil
+}
+
+// applyRules generates new mappings based on the given set of rules
+func (dbm *databaseMappings) applyRules(rules *schema.ClusterMappingRuleSet) error {
+	dbm.log.Infof("applying rule set %d for database %s", rules.ForVersion, dbm.name)
+
+	dbm.log.Infof("applying %d transitions for database %s", len(rules.ShardTransitions), dbm.name)
+	for _, t := range rules.ShardTransitions {
+		ruleShards := bitset.From(t.Shards.Bits)
+
+		// if there is no FromCluster, this is an initial assignment
+		if t.FromCluster == "" {
+			dbm.log.Infof("assigning %d shards to %s", ruleShards.Count(), t.ToCluster)
+			dbm.active = append(dbm.active, &activeClusterMapping{
+				shards: ruleShards,
+				clusterMapping: clusterMapping{
+					cluster:          t.ToCluster,
+					readCutoverTime:  xtime.FromUnixMillis(t.ReadCutoverTime),
+					writeCutoverTime: xtime.FromUnixMillis(t.WriteCutoverTime),
+				},
+			})
+			continue
+		}
+
+		// transition from prior clusters
+		for _, a := range dbm.active {
+			shards := a.shards.Intersection(ruleShards)
+			if shards.None() {
+				continue
+			}
+
+			if a.cluster != t.FromCluster {
+				// Gah! Something is broken
+				dbm.log.Errorf(`expecting to transition %d shards from cluster %s to 
+cluster %s but those shards are currently assigned to %s`,
+					shards.Count(), t.FromCluster, t.ToCluster, a.cluster)
+				return errInvalidTransition
+			}
+
+			// create a mapping that closes the currently active mapping, then create
+			// a new active mapping pointing to the new cluster
+			dbm.log.Infof("transitioning %d shards from cluster %s:%s to %s:%s",
+				shards.Count(), dbm.name, t.FromCluster, dbm.name, t.ToCluster)
+			dbm.active = append(dbm.active, &activeClusterMapping{
+				shards: shards,
+				clusterMapping: clusterMapping{
+					cluster:          t.ToCluster,
+					readCutoverTime:  xtime.FromUnixMillis(t.ReadCutoverTime),
+					writeCutoverTime: xtime.FromUnixMillis(t.WriteCutoverTime),
+					prior: &clusterMapping{
+						cluster:          a.cluster,
+						readCutoverTime:  a.readCutoverTime,
+						writeCutoverTime: a.writeCutoverTime,
+						cutoffTime:       xtime.FromUnixMillis(t.CutoverCompleteTime),
+						prior:            a.prior,
+					},
+				},
+			})
```

Comment:
Same comment as above; I'd prefer to see the code inline rather than have a bunch of one-off constructors.